fdbclient/include/fdbclient/MetaclusterManagement.actor.h

```diff
@@ -1152,31 +1153,46 @@ struct CreateTenantImpl {
 
 			if (groupEntry.present()) {
 				ASSERT(groupEntry.get().assignedCluster.present());
+				if (self->preferAssignedCluster) {
+					ASSERT(groupEntry.get().assignedCluster.get() == self->tenantEntry.assignedCluster.get());
+				}
 				return std::make_pair(groupEntry.get().assignedCluster.get(), true);
 			}
 		}
 
 		// Get a set of the most full clusters that still have capacity
-		state KeyBackedSet<Tuple>::RangeResultType availableClusters =
-		    wait(ManagementClusterMetadata::clusterCapacityIndex.getRange(
-		        tr, {}, {}, CLIENT_KNOBS->METACLUSTER_ASSIGNMENT_CLUSTERS_TO_CHECK, Snapshot::False, Reverse::True));
-
-		if (availableClusters.results.empty()) {
-			throw metacluster_no_capacity();
-		}
-
+		// If preferred cluster is specified, look for that one.
 		state std::vector<Future<Reference<IDatabase>>> dataClusterDbs;
-		for (auto clusterTuple : availableClusters.results) {
-			dataClusterDbs.push_back(getAndOpenDatabase(tr, clusterTuple.getString(1)));
-		}
-
-		wait(waitForAll(dataClusterDbs));
-
-		// Check the availability of our set of clusters
 		state std::vector<Future<ClusterName>> clusterAvailabilityChecks;
-		for (int i = 0; i < availableClusters.results.size(); ++i) {
+		if (self->preferAssignedCluster) {
+			dataClusterDbs.push_back(getAndOpenDatabase(tr, self->tenantEntry.assignedCluster.get()));
+			wait(waitForAll(dataClusterDbs));
 			clusterAvailabilityChecks.push_back(
-			    checkClusterAvailability(dataClusterDbs[i].get(), availableClusters.results[i].getString(1)));
+			    checkClusterAvailability(dataClusterDbs[0].get(), self->tenantEntry.assignedCluster.get()));
+		} else {
+			state KeyBackedSet<Tuple>::RangeResultType availableClusters =
+			    wait(ManagementClusterMetadata::clusterCapacityIndex.getRange(
+			        tr,
+			        {},
+			        {},
+			        CLIENT_KNOBS->METACLUSTER_ASSIGNMENT_CLUSTERS_TO_CHECK,
+			        Snapshot::False,
+			        Reverse::True));
+
+			if (availableClusters.results.empty()) {
+				throw metacluster_no_capacity();
+			}
+
+			for (auto clusterTuple : availableClusters.results) {
+				dataClusterDbs.push_back(getAndOpenDatabase(tr, clusterTuple.getString(1)));
+			}
+			wait(waitForAll(dataClusterDbs));
+
+			// Check the availability of our set of clusters
+			for (int i = 0; i < availableClusters.results.size(); ++i) {
+				clusterAvailabilityChecks.push_back(
+				    checkClusterAvailability(dataClusterDbs[i].get(), availableClusters.results[i].getString(1)));
+			}
```

Comment:
I think this works fine. If you wanted you could structure this slightly different where we collect a vector of cluster names to check, and then the `if/else` block would get cheaper and you could avoid duplicating the logic to do the checks.